services/worker/src/worker/job_runners/split/descriptive_statistics.py

```diff
@@ -0,0 +1,412 @@
+    n_bins: int,
+    n_samples: int,
+    column_type: ColumnType,
+) -> NumericalStatisticsItem:
+    logging.debug(f"Compute min, max, mean, median, std and proportion of null values for {column_name}")
+    min_max_mean_median_std_command = COMPUTE_MIN_MAX_MEAN_MEDIAN_STD_COMMAND.format(
+        column_name=column_name, parquet_filename=parquet_filename
+    )
+    minimum, maximum, mean, median, std = con.sql(min_max_mean_median_std_command).fetchall()[0]
+    logging.debug(f"{minimum=}, {maximum=}, {mean=}, {median=}, {std=}")
+
+    nan_count_command = COMPUTE_NAN_COUNTS_COMMAND.format(column_name=column_name, parquet_filename=parquet_filename)
+    nan_count = con.sql(nan_count_command).fetchall()[0][0]
+    nan_proportion = np.round(nan_count / n_samples, DECIMALS).item() if nan_count else 0.0
+    logging.debug(f"{nan_count=} {nan_proportion=}")
+    histogram = compute_histogram(
+        con,
+        column_name,
+        parquet_filename,
+        min_value=minimum,
+        max_value=maximum,
+        column_type=column_type,
+        n_bins=n_bins,
+        n_samples=n_samples - nan_count,
+    )
+    if column_type == ColumnType.FLOAT:
+        minimum, maximum, mean, median, std = np.round([minimum, maximum, mean, median, std], DECIMALS).tolist()
+    elif column_type == ColumnType.INT:
+        mean, median, std = np.round([mean, median, std], DECIMALS).tolist()
+    else:
+        raise ValueError(f"Incorrect column type {column_type}")
+    return NumericalStatisticsItem(
+        nan_count=nan_count,
+        nan_proportion=nan_proportion,
+        min=minimum,
+        max=maximum,
+        mean=mean,
+        median=median,
+        std=std,
+        histogram=histogram,
+    )
+
+
+def compute_categorical_statistics(
+    con: duckdb.DuckDBPyConnection,
+    column_name: str,
+    parquet_filename: Path,
+    class_label_names: List[str],
+    n_samples: int,
+) -> CategoricalStatisticsItem:
+    categorical_counts_query = COMPUTE_CATEGORIES_COUNTS_COMMAND.format(
+        column_name=column_name, parquet_filename=parquet_filename
+    )
+    categories: List[Tuple[int, int]] = con.sql(
+        categorical_counts_query
+    ).fetchall()  # list of tuples (idx, num_samples)
+
+    logging.debug(f"Statistics for {column_name} computed")
+    frequencies, nan_count = {}, 0
+    for cat_id, freq in categories:
+        if cat_id is not None:
+            frequencies[class_label_names[cat_id]] = freq
+        else:
+            nan_count = freq
+    nan_proportion = np.round(nan_count / n_samples, DECIMALS).item() if nan_count != 0 else 0.0
+    return CategoricalStatisticsItem(
+        nan_count=nan_count,
+        nan_proportion=nan_proportion,
+        n_unique=len(categories) - 1 if nan_count else len(categories),
+        frequencies=frequencies,
+    )
+
+
+def compute_descriptive_statistics_response(
+    dataset: str,
+    config: str,
+    split: str,
+    local_parquet_directory: Optional[Path],
```

Comment:
If I understand correctly it doesn't need to be Optional, since the SplitDescriptiveStatisticsJobRunner always provides the local_parquet_directory

Reply:
this is because `cache_subdirectory` attribute of `JobRunnerWithCache` is [typed as `Optional[Path]`](https://github.com/huggingface/datasets-server/blob/main/services/worker/src/worker/job_runners/_job_runner_with_cache.py#L23), not sure what to do, here it really doesn't look logical I agree.

Reply:
i mean, `mypy` gives error if make it just `Path`, I can put `type: ignore` ofc

Reply:
Yes, it should not be optional. See https://github.com/huggingface/datasets-server/blob/0691cb7af3cd3a2293ad0120c684bb686f7aa804/services/worker/src/worker/job_runners/split/duckdb_index.py#L281-L282